cpp/src/arrow/compute/row/encode_internal.cc

```diff
@@ -23,9 +23,10 @@ namespace compute {
 
 void RowTableEncoder::Init(const std::vector<KeyColumnMetadata>& cols, int row_alignment,
                            int string_alignment) {
-  row_metadata_.FromColumnMetadataVector(cols, row_alignment, string_alignment);
-  uint32_t num_cols = row_metadata_.num_cols();
-  uint32_t num_varbinary_cols = row_metadata_.num_varbinary_cols();
+  row_metadata_ = std::make_shared<RowTableMetadata>();
```

Comment:
Mainly modified here , initializes a row_metadata_'s pointer and passes it to RowTableImpl